### PR TITLE
Feature/assisted steering

### PR DESCRIPTION
**Pull request by KianChilds**
## Justification
Makes steering simpler

## Implementaion
A button that activates auto rotation. Using inference to determine which target to rotate to.

## Testing Done
as of 3/6/24 none

### CODE REVIEW

**Comment by KianChilds** [comment]
fixes issue #195

File: src/main/java/frc/robot/Constants.java [hunk 1]
```diff
@@ -295,4 +297,12 @@ public static class enabledSubsystems {
...
   // Lights
   public static final int LED_COUNT = 20;
+
+  public static class DriverConstants {
+
+    public static final Rotation2d RED_AMP_ROTATION = null;
+    public static final Rotation2d BLUE_AMP_ROTATION = null;
+    public static final Rotation2d BLUE_SOURCE_ROTATION = null;
+    public static final Rotation2d RED_SOURCE_ROTATION = null;
+  }
```
Comment (azvanderpas) [inline src/main/java/frc/robot/Constants.java:307]: Seem like these fit in FieldConstants. Also can't leave null, of course

File: src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java [hunk 2]
```diff
@@ -190,4 +195,77 @@ public Command zeroDriveTrain() {
...
+                autoTargetSpeaker(requestSupplier, RSM),
+                isAmpMode),
+            Commands.either(Commands.none(), autoTargetSource(requestSupplier, RSM), isAmpMode),
+            onNearSide);
+
+    return assistDriver;
+  }
+
+  private BooleanSupplier isOnNearSide() {
+    return () -> subsystem.getState().Pose.getX() > FieldConstants.CENTERLINE_X;
```
Comment (azvanderpas) [inline src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java:245]: Isn't this alliance color dependent as well?

File: src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java [hunk 3]
```diff
@@ -190,4 +195,77 @@ public Command zeroDriveTrain() {
...
+  }
+
+  private Command autoTargetSource(Supplier<DriveRequest> request, RobotStateManager RSM) {
+    return Commands.either(
+            pointInDirection(DriverConstants.RED_SOURCE_ROTATION, request),
+            pointInDirection(DriverConstants.BLUE_SOURCE_ROTATION, request),
+            () -> RSM.getAllianceColor() == AllianceColor.RED)
+        .asProxy();
```
Comment (azvanderpas) [inline src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java:253]: Add withNames

File: src/main/java/frc/robot/Constants.java [hunk 4]
```diff
@@ -272,6 +273,7 @@ public static class ClimberConstants {
   public static class FieldConstants {
     public static final Translation2d RED_SPEAKER = new Translation2d(16.579342, 5.547868);
     public static final Translation2d BLUE_SPEAKER = new Translation2d(-0.0381, 5.547868);
+    public static final double CENTERLINE_X = 0;
```
Comment (azvanderpas) [inline src/main/java/frc/robot/Constants.java:276]: Is this a TODO number?

**Review by azvanderpas — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 5]
```diff
@@ -139,6 +142,9 @@ public void teleopInit() {
     if (m_autonomousCommand != null) {
       m_autonomousCommand.cancel();
     }
+    if (DriverStation.getAlliance().get() == Alliance.Red) {
```
Comment (lohengrin332) [inline src/main/java/frc/robot/Robot.java:145]: A comment might be nice here. E.g.:

```suggestion
    // BLUE is the default, so we only need to handle RED here.
    if (DriverStation.getAlliance().get() == Alliance.Red) {
```

File: src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -242,17 +245,17 @@ private void configureBindings() {
...
-        .whileTrue(
-            Commands.startEnd(
-                () -> OI.Driver.setRumble(Constants.OperatorConstants.RUMBLE_STRENGTH),
-                () -> OI.Driver.setRumble(0)));
-    new Trigger(shooterSubsystem::isShooterReady)
-        .whileTrue(
-            Commands.startEnd(
-                () -> OI.Operator.setRumble(Constants.OperatorConstants.RUMBLE_STRENGTH),
-                () -> OI.Operator.setRumble(0)));
+    // new Trigger(trapElvCommandFactory.getSourceBreak())
```
Comment (lohengrin332) [inline src/main/java/frc/robot/RobotContainer.java:248]: Per our conversation, please un-comment this block.

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -261,14 +264,16 @@ private Command speakerSource() {
   }
 
   private Command intakeSpeaker() {
-    return Commands.parallel(
-        shooterCommandFactory.intakeSpeakerSource(), triggerCommandFactory.getLoadCommand());
+    return Commands.none();
+    // return Commands.parallel(
```
Comment (lohengrin332) [inline src/main/java/frc/robot/RobotContainer.java:268]: Please also uncomment this block.

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -261,14 +264,16 @@ private Command speakerSource() {
...
+    //     intakeCommandFactory.intakeSpeakerCommandSmart(shooterSubsystem.getBeamBreak()),
+    //     triggerCommandFactory.getGroundLoadCommand(shooterSubsystem.getBeamBreak()));
   }
 
   private Command intakeAmp() {
     return Commands.parallel(
             trapElvCommandFactory.intakeGround(), intakeCommandFactory.getAmpIntakeCommand())
-        .until(trapElvCommandFactory.getSourceBreak())
+        // .until(trapElvCommandFactory.getSourceBreak())
```
Comment (lohengrin332) [inline src/main/java/frc/robot/RobotContainer.java:276]: Please also uncomment this block.

File: src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java [hunk 9]
```diff
@@ -190,4 +195,85 @@ public Command zeroDriveTrain() {
...
+     *  yes{
+     *   no op
+     *  }
+     *  no{
+     *   target source
+     *  }
+     * }
+     */
+
+    BooleanSupplier onNearSide = isOnNearSide(() -> RSM.getAllianceColor() == AllianceColor.RED);
```
Comment (lohengrin332) [inline src/main/java/frc/robot/subsystems/swerveSubsystem/SwerveCommandFactory.java:229]: These should be made final.

**Review by lohengrin332 — APPROVED** [review]
LGTM

**Review by lohengrin332 — APPROVED** [review]

**Pull request merged**